components/sup/src/http_gateway.rs

```diff
@@ -112,71 +115,93 @@ impl AppState {
         }
-    }
-}
-
-struct Metrics;
+    };
 
-impl Middleware<AppState> for Metrics {
-    fn start(&self, req: &HttpRequest<AppState>) -> actix_web::Result<Started> {
-        let label_values = &[req.path()];
+    // From this point forward, we know that we have an
+    // authentication token in the state. Therefore, anything
+    // short of a fully formed Authorization header (yes,
+    // Authorization; HTTP is fun, kids!) containing a Bearer
+    // token that matches the value we have in our state, results
+    // in an Unauthorized response.
+    let hdr = match req.headers()
+                       .get(http::header::AUTHORIZATION)
+                       .ok_or("header missing")
+                       .and_then(|hv| hv.to_str().or(Err("can't convert to str")))
+    {
+        Ok(h) => h,
+        Err(e) => {
+            debug!("Error reading required Authorization header: {:?}.", e);
+            return Either::B(ok(req.into_response(HttpResponse::Unauthorized().finish())));
+        }
+    };
 
-        HTTP_GATEWAY_REQUESTS.with_label_values(label_values).inc();
-        let timer = HTTP_GATEWAY_REQUEST_DURATION.with_label_values(label_values)
-                                                 .start_timer();
-        req.state().timer.set(Some(timer));
+    let hdr_components: Vec<&str> = hdr.split_whitespace().collect();
 
-        Ok(Started::Done)
+    match hdr_components.as_slice() {
+        ["Bearer", incoming_token] if crypto::secure_eq(current_token, incoming_token) => {
+            Either::A(srv.call(req))
+        }
+        _ => Either::B(ok(req.into_response(HttpResponse::Unauthorized().finish()))),
     }
+}
 
-    fn finish(&self, req: &HttpRequest<AppState>, _resp: &HttpResponse) -> Finished {
-        let timer = req.state().timer.replace(None);
-
-        if timer.is_some() {
-            timer.unwrap().observe_duration();
-        }
+fn metrics_middleware<S>(req: ServiceRequest,
+                         srv: &mut S)
+                         -> impl Future<Item = ServiceResponse<Body>, Error = Error>
+    where S: Service<Request = ServiceRequest, Response = ServiceResponse<Body>, Error = Error>
+{
+    let label_values = &[req.path()];
+
+    HTTP_GATEWAY_REQUESTS.with_label_values(label_values).inc();
+    let timer = HTTP_GATEWAY_REQUEST_DURATION.with_label_values(label_values)
+                                             .start_timer();
+    req.app_data::<AppState>()
+       .expect("app data")
+       .timer
+       .set(Some(timer));
+
+    srv.call(req).and_then(|res| {
+                     let timer = res.request()
+                                    .app_data::<AppState>()
+                                    .expect("app data")
+                                    .timer
+                                    .replace(None);
+                     if timer.is_some() {
+                         timer.unwrap().observe_duration();
+                     }
```

Comment:
I realize this is how the code was already, but it seems simpler/idiomatic (and avoids an unwrap) to:
```rust
                     if let Some(timer) = res.request()
                                             .app_data::<AppState>()
                                             .expect("app data")
                                             .timer
                                             .replace(None)
                     {
                         timer.observe_duration();
                     }
```